talk/expert/cpp20spaceship.tex

```diff
@@ -0,0 +1,242 @@
+    \begin{itemize}
+      \item The type {\it std::strong_ordering} is convertible to {\it std::partial_ordering}, but not the contrary.
+      \item Between the two, we have a class {\it std::weak_ordering}, where {\it ==0} mean that the two compared values are equivalent from a ranking point of view, but not necessarily \textbf{equal}. In a given expression, one cannot substitutes one value for the other and be sure to have the same result.
+      \item I am not aware of some predefined type whose {\it <=>} would return an instance of {\it std::weak_ordering}, but it may make sense for some home-made class, such as the following.
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Home-made class}
+    \begin{itemize}
+    \item In the example below, we define a very basic class for positive rational numbers, and provide an implementation of {\it <=>}.
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    \begin{cppcode*}{}
+#include <iostream>
+#include <compare>
+
+struct Rational
+ {
+  unsigned n, d ;
+
+  std::weak_ordering operator<=>( Rational const & other ) const
+   { return (n*other.d)<=>(other.n*d) ; }
+ } ;
+
+std::ostream & operator<<( std::ostream & os, std::weak_ordering cmp )
+ { return (os<<'<'<<(cmp<0)<<'|'<<(cmp==0)<<'|'<<(cmp>0)<<'>') ; }
+
+std::ostream & operator<<( std::ostream & os, Rational const & r )
+ { return (os<<r.n<<'/'<<r.d) ; }
+
+template <typename T>
+void compare( T lhs, T rhs )
+ {
+  std::cout<<std::endl ;
+  //std::cout<<lhs<<" ==  "<<rhs<<": "<<(lhs==rhs)<<std::endl ;
+  //std::cout<<lhs<<" !=  "<<rhs<<": "<<(lhs!=rhs)<<std::endl ;
+  std::cout<<lhs<<" <=> "<<rhs<<": "<<(lhs<=>rhs)<<std::endl ;
+  std::cout<<lhs<<" <   "<<rhs<<": "<<(lhs<  rhs)<<std::endl ;
+  std::cout<<lhs<<" <=  "<<rhs<<": "<<(lhs<= rhs)<<std::endl ;
+  std::cout<<lhs<<" >   "<<rhs<<": "<<(lhs>  rhs)<<std::endl ;
+  std::cout<<lhs<<" >=  "<<rhs<<": "<<(lhs>= rhs)<<std::endl ;
+ }
+
+int main()
+ {
+  compare<Rational>({ 3, 4 },{ 2, 3 }) ;
+  compare<Rational>({ 3, 6 },{ 2, 3 }) ;
+  compare<Rational>({ 1, 2 },{ 2, 4 }) ;
+ }
+     \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Home-made class}
+    \begin{itemize}
+      \item Despites the {\it <=>} for {\it unsigned} returns some instance of {\it std::strong_ordering}, we prefer here to cast it to {\it std::weak_ordering}. This way, we emphasize that if {\it a<=>b} is equal to {\it 0}, it only means that {\it a} and {\it b} are logically equivalent, but may lead to different results in some expressions.
+      \item As one can check in the code, we have defined {\it <=>}, and the compiler has added, for free, some default implementations for {\it <}, {\it >}, {\it <=} and {\it =>}. Of course, one can overload them with home-made implementations.
```

Comment:
This is technically not correct, because the compiler does not provide the other operators. It rewrites their occurance in terms of `<=>`.
```suggestion
      \item As one can see from the code, we have defined only {\it <=>}, but {\it <}, {\it >}, {\it <=} and {\it =>} work as well. The compiler can express the latter operators in terms of {\it <=>}. Of course, one can also provide own implementations.
```

Reply:
Thanks for this desambiguation. Actually, that was a point to be investigated in my TODO, because I read contradictory statements about this.